design/0002-ci-health-metrics.md

```diff
@@ -0,0 +1,154 @@
+- Reviewers: @grampelberg
+
+# Summary
+
+[summary]: #summary
+
+This RFC proposes a service that would monitor the results from each CI run. It
+should allow us to catch consistently flaky tests.
+
+# Problem Statement (Step 1)
+
+[problem-statement]: #problem-statement
+
+CI jobs inevitably fail occasionally, either because of real problems with the
+code, or because of test flakiness. We'd like to stay informed about what tests
+are being persistently flaky, without having to dig into the CI run results
+themselves. This will allow us to properly prioritize work to fix those
+failures.
+
+More specifically we'd like to get, for a given time span, the list of error
+messages with their frequency aggregated globally, per workflow and per job.
+
+The first problem to address will be to expose the CI jobs results, which
+currently are only available through Github's UI. With access to that data in
+the appropriate format, we'll then be able to determine how to persist and
+present it to operators.
+
+# Design proposal (Step 2)
+
+## Exposing the errors
+
+There are two ways to expose CI run failures: annotations and artifacts.
+
+### Annotations
+
+Annotations are created whenever an error message from any of the CI steps is
+sent to stdout/stderr using the appropriate format, described
+[here](https://help.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-error-message).
+Annotation messages are visible in the summary page of a CI run. Currently our
+error messages aren't following that format, and the summary page only shows
+something like "Process completed with exit code 1." whenever a job fails.
+Annotations are surfaced through Github's Checks API.
+
+### Artifacts
+
+Instead of conforming to Github's annotation format, we could use a more
+standard format such as xUnit XML reports. That would imply building some
+wrapper around the tests to accommodate that format, and the results would be
+then uploaded to the CI run's artifact repository (which we're already
+leveraging to store docker images). Then another tool can consume those reports.
+
+But I haven't found such a tool that would consume xUnit reports for showing
+historical trends, which is the main goal of this RFC (being able to catch
+consistently flaky tests). For this reason I'm proposing we go down the
+annotations route.
+
+(Although do raise your voice if you find an appropriate tool for xUnit
+reports!)
+
+## Formatting
+
+We have integration tests and checks written in go, javascript and bash. Each of
+those environments would require a wrapper around their output to properly
+format errors as explained above. The error message would need to have at least
+two elements: A generic message that can be aggregated in a counter (see below)
+and a message with more details about this particular run.
+
+## Architecture
+
+The plan being to ultimately present the historical trend of errors, the obvious
```

Comment:
To me, just a regular (not timeseries) database seems like a better fit here.  It lets us have columns for things like test name, source file, error message, line number, test duration, timestamp, etc. without needing to worry about making all of these fields into Prometheus labels and handling the cardinality.

This easily lets us do queries like:
* list all of the tests sorted by success rate over the last 30 days
* for a given test, list the most recent 10 failures and their error messages

This feels a lot more like a database of events and less like a timeseries.

Reply:
Either way, it would be helpful to list the columns or labels that we plan to record.

Reply:
The sole fact these are data points collected over time, led me to suggest a TSDB. I wasn't worried by cardinality given the few amount of data this will handle. I'll have to confess though I don't know how easy it'd be to manage that data as opposed to simple SQL queries in a regular database, so I'm willing to reconsider.

Reply:
If we plan to put the detailed error messages as Prometheus labels, I would assume these are all unique and we would get each data point alone in its own timeseries, which is kind of a weird way to use Prometheus.